include/nbl/asset/ICPURayTracingPipeline.h

```diff
@@ -0,0 +1,128 @@
+                core::vector<SShaderSpecInfo> results;
+                results.resize(specInfos.size());
+                for (auto specInfo_i = 0u; specInfo_i < specInfos.size(); specInfo_i++)
+                    results[specInfo_i] = specInfos[specInfo_i].clone(depth);
+                return results;
+            };
+            newPipeline->m_misses = cloneSpecInfos(m_misses);
+            newPipeline->m_hitGroups.anyHits = cloneSpecInfos(m_hitGroups.anyHits);
+            newPipeline->m_hitGroups.closestHits = cloneSpecInfos(m_hitGroups.closestHits);
+            newPipeline->m_hitGroups.intersections = cloneSpecInfos(m_hitGroups.intersections);
+            newPipeline->m_callables = cloneSpecInfos(m_callables);
+
+            newPipeline->m_params = m_params;
+            return core::smart_refctd_ptr<base_t>(newPipeline);
+        }
+
+        constexpr static inline auto AssetType = ET_RAYTRACING_PIPELINE;
+        inline E_TYPE getAssetType() const override { return AssetType; }
+        
+        //!
+        inline size_t getDependantCount() const override { 
+            //TODO(kevinyu): Remove this function use computeDependants
+            return 0;
+        }
+
+        virtual core::unordered_set<const IAsset*> computeDependants() const override final {
+            core::unordered_set<const IAsset*> dependants;
+            dependants.insert(m_raygen.shader.get());
+            for (const auto& missInfo : m_misses) dependants.insert(missInfo.shader.get());
+            for (const auto& anyHitInfo : m_hitGroups.anyHits) dependants.insert(anyHitInfo.shader.get());
+            for (const auto& closestHitInfo : m_hitGroups.closestHits) dependants.insert(closestHitInfo.shader.get());
+            for (const auto& intersectionInfo : m_hitGroups.intersections) dependants.insert(intersectionInfo.shader.get());
+            for (const auto& callableInfo : m_callables) dependants.insert(callableInfo.shader.get());
+            return dependants;
+        }
+
+        inline virtual std::span<const SShaderSpecInfo> getSpecInfo(hlsl::ShaderStage stage) const override final
+        {
+            switch (stage) 
+            {
+                case hlsl::ShaderStage::ESS_RAYGEN:
+                  return { &m_raygen, 1 };
+                case hlsl::ShaderStage::ESS_MISS:
+                  return m_misses;
+                case hlsl::ShaderStage::ESS_ANY_HIT:
+                  return m_hitGroups.anyHits;
+                case hlsl::ShaderStage::ESS_CLOSEST_HIT:
+                  return m_hitGroups.closestHits;
+                case hlsl::ShaderStage::ESS_INTERSECTION:
+                  return m_hitGroups.intersections;
+                case hlsl::ShaderStage::ESS_CALLABLE:
+                  return m_callables;
+
+            }
+            return {};
+        }
+
+        inline virtual bool valid() const override final
+        {
+            // TODO(kevinyu): Fix this temporary dummy code
+            return true;
```

Comment:
If we're on the same page, then this is baiscally checking a layout and raygen shaders are there

All other stages are optional, but if they are present, they need to be checked for validity (spec constants and stuff)

Remember that we can patch NO_NULL flags in the asset converter if we detect incompatibility (missing stages), those will be in our patch.